### DISCUSSION TITLE
eUPF's eBPF code for gNodeB simulation tool

### DISCUSSION THREAD

**Discussion by linouxis9** [edgecomllc/eupf#469] [General]
Hi,

I'm working on a (private for now) fork of my5G-RANTester (https://github.com/my5G/my5G-RANTester) which is a tool similar to UERANSIM. My fork is able to simulate multiple UEs (10k+) and gNodeB and do performance testing of the control plane.
My goal is to improve my fork in same the vein to be able to do performance testing of the user plane with it as well.
I'm currently using https://github.com/free5gc/gtp5g as the GTP-U stack inside my tool for this job, the performance is great, but it has a bunch of shortcomings which using your eBPF code would fix.
As such, I'm looking to replace free5gc's gtp5g using your eUPF's eBPF code for that task. Would you recommend it? Would it fit well a RAN simulator?
The goal is to have a simulator (which can simulate tons of UEs and gNodeBs from a single tool) with an user plane much more performant than UERANSIM for instance.

Thanks a lot!

**Comment by pirog-spb** [comment]
Hi Valentin,

Feel free to use ebpf part of eUPF if it helps :)

BTW we are interested to test/improve performance of our eUPF. So if you could share results or even share your tool - it would be very nice.

BR. Alex

**Reply by linouxis9** [reply]
Hi Alex,

Thanks for the positive reply! I'm trying to use your eBPF programs, and will report issues that I may find with it.
If I manage to get the tool in a decent shape, I'll be sure to release it open-sourcely as well, and ping you about it. The open-source community is definitely lacking of an UERANSIM-like tool but focused on performance.

Thanks and cheers,
Valentin

**Reply by linouxis9** [reply]
Hi Alex @pirog-spb,

I'm proud to announce that my tool is now open-source! [PacketRusher](https://github.com/HewlettPackard/PacketRusher) is an high performance 5G UE/gNB Simulator and CP/UP load tester.
For now, we are still using the legacy free5gc's gtp5g module (I'll release my eBPF test in another branch soon), so some limitations may apply, and documentation is still lacking for now.
Feel free to test it with eUPF, and if you have any issues using the tool, feel free to reach out to me!

Thanks and cheers,
Valentin

**Reply by pirog-spb** [reply]
Hi Valentin @linouxis9 

Great news! We will definitely try your tool. 
Actually, we are experiencing troubles with gNBSim from OMEC project and looking for alternatives.

**Reply by linouxis9** [reply]
Hi Alex @pirog-spb,
Glad to hear! Feel free to reach out to me if you need any help or information.
I hope we can collaborate together on the eBPF code as well, especially for its integration inside PacketRusher :-)
Thanks and cheers,
Valentin

**Reply by linouxis9** [reply]
Hi Alex @pirog-spb,
I finally got the time to continue on my PacketRusher experiments using eUPF's eBPF code!
I'm successfully able to do ICMP traffic between simulated UEs and VMs on the data network behind the UPF.
However, there is a small issue that is leaving me scratching my head: while ICMP traffic work flawlessly, VMs on the data network never reply to TCP packets sent by the UEs. Using tcpdump, we can observe that the TCP SYN packet is correctly received on Data Network's VMs, but TCP SYN-ACK is never sent.
Is it an issue you observed while developing eUPF, or would you know what could be wrong?

**EDIT: I was finally able to make it working! I had to disable TCP checksum offloading on the UE's interface as the TCP checksum was wrong, I guess the encapsulation somehow messes the TCP's checksum.**

Thanks a lot for the eUPF!!
Valentin
